### ISSUE TITLE
[X86] Runtime library does not utilize AVX2 and AVX512

### ISSUE THREAD

**Issue by ZviRackover** [halide/Halide#1628]
The X86 runtime library should be updated with functions that utilize:

1. New instructions added in AVX2 and AVX512
2. Wider vectors for AVX512

**Comment by abadams** [comment]
Are you thinking of things like floor, ceil, round, sqrt, etc? A good start might be copy-pasting src/runtime/x86_avx.ll and doubling all the vector widths.

**Comment by ZviRackover** [comment]
Yes. That's what I had in mind. Do we have any data on how important is this library for performance of typical applications?

**Comment by abadams** [comment]
Quite a lot of floating point code uses those functions, so I'm guessing
implementing AVX512-width ones would be a minor benefit on those platforms.

**Comment by ZviRackover** [comment]
I started implementing abs_i16x16 which should be lowered to
`vpabsw	%ymm0, %ymm0`

It turns out that the following implementation is lowered to the above instruction. There is no need to call an X86 intrinsics function. This works with LLVM top of trunk (the next release is 4.0), but I'm not sure about earlier releases.

```
define weak_odr <16 x i16> @abs_i16x16(<16 x i16> %x) nounwind uwtable readnone alwaysinline {
   %cmp = icmp sgt <16 x i16> %x, <i16 -1, i16 -1, i16 -1, i16 -1, i16 -1, i16 -1, i16 -1, i16 -1, i16 -1, i16 -1, i16 -1, i16 -1, i16 -1, i16 -1, i16 -1, i16 -1>
   %neg = sub <16 x i16> zeroinitializer, %x
   %res = select <16 x i1> %cmp, <16 x i16> %x, <16 x i16> %neg
   ret <16 x i16> %res
 }
```

Another possibility may be instead of implementing the above function, change CodeGen_X86::visit(Call*) to generate a vector select(x > -1 ? x : -x). Which should be lowered to the same IR as the above implementation. Does this sound like a better solution in terms of code maintainability?
Sorry if this question doesn't make sense. I am a newb to Halide.

**Comment by abadams** [comment]
If no abs intrinsic is found in the runtime, we generate select(x >= 0, x,
-x)

If that generates the right code on all x86 flavors on all recent llvm
versions (>3.7), then we can just delete all the abs intrinsics for X86.
That would be nice. We can also conditionalize them on an llvm version
(e.g. ifdef LLVM_VERSION < 40)

**Comment by ZviRackover** [comment]
Ok, I will try to remove all abs_i* X86 flavors and see if we still get the optimal instruction selection. I would expect this to happen since the LLVM X86 backend supports these patterns. [See the relevant tests for reference.](https://github.com/llvm-mirror/llvm/blob/master/test/CodeGen/X86/viabs.ll)

**Comment by dsharletg** [comment]
I just tried removing our handling of abs and it still doesn't generate the right instructions now.
